### PR TITLE
noderesourcetopology: add integration tests for scope=container

### PR DESCRIPTION
**Pull request by Tal-or**
#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:

Add a batch of tests that are validating the flow of NodeResourceTopology plugin when Kubelet is configured with TopologyManagerScope=container.

Each test has a type and tier level which are correspond to the table in the pkg/noderesourcetopology/TESTS.md file.

#### Special notes for your reviewer:
This PR depends and based on #371

### CODE REVIEW

**Comment by Tal-or** [comment]
/hold
depends on #371

File: test/integration/noderesourcetopology_test.go [hunk 1]
```diff
@@ -555,6 +559,286 @@ func TestTopologyMatchPlugin(t *testing.T) {
...
+						{
+							noderesourcetopology.MakeTopologyResInfo(cpu, "4", "4"),
+							noderesourcetopology.MakeTopologyResInfo(memory, "8Gi", "8Gi"),
+						},
+					}),
+			},
+			expectedNodes: []string{"fake-node-2"},
+		},
+		{
+			name: "SingleNUMANodeContainerLevel: Filtering out nodes that cannot fit resources in case of Guaranteed pod with init container",
```
Comment (fromanirh) [inline test/integration/noderesourcetopology_test.go:654]: same as above.

File: test/integration/noderesourcetopology_test.go [hunk 2]
```diff
@@ -555,6 +559,286 @@ func TestTopologyMatchPlugin(t *testing.T) {
...
+						{
+							noderesourcetopology.MakeTopologyResInfo(cpu, "4", "4"),
+							noderesourcetopology.MakeTopologyResInfo(memory, "8Gi", "8Gi"),
+						},
+					}),
+			},
+			expectedNodes: []string{"fake-node-2"},
+		},
+		{
+			name: "SingleNUMANodeContainerLevel: Cannot fit resources in case of Guaranteed pod with multi containers",
```
Comment (fromanirh) [inline test/integration/noderesourcetopology_test.go:689]: same as above - if possible. We should try to have tests on which `sum(R) < Node` - if this is not possible, or if we can exercise a corner case, we can do otherwise, but let's add comments to explain why we add this exception.

File: test/integration/noderesourcetopology_test.go [hunk 3]
```diff
@@ -555,6 +559,286 @@ func TestTopologyMatchPlugin(t *testing.T) {
...
+			expectedNodes: []string{"fake-node-1"},
+		},
+		{
+			name: "Negative: SingleNUMANodeContainerLevel: Cannot fit resources in case of Guaranteed pod with init container",
+			pods: []*v1.Pod{
+				initWithReqAndLimit(
+					withReqAndLimit(
+						st.MakePod().Namespace(ns).Name("topology-aware-scheduler-pod"),
+						map[v1.ResourceName]string{v1.ResourceCPU: "2", v1.ResourceMemory: "4Gi"}),
+					map[v1.ResourceName]string{v1.ResourceCPU: "4", v1.ResourceMemory: "10Gi"}).Obj(),
```
Comment (fromanirh) [inline test/integration/noderesourcetopology_test.go]: OTOH this test seems fine!

File: test/integration/noderesourcetopology_test.go [hunk 4]
```diff
@@ -555,6 +559,286 @@ func TestTopologyMatchPlugin(t *testing.T) {
...
+			expectedNodes: []string{},
+		},
+		{
+			name: "Negative: SingleNUMANodeContainerLevel: Cannot fit resources in case of Guaranteed pod with init container",
+			pods: []*v1.Pod{
+				initWithReqAndLimit(
+					withReqAndLimit(
+						st.MakePod().Namespace(ns).Name("topology-aware-scheduler-pod"),
+						map[v1.ResourceName]string{v1.ResourceCPU: "2", v1.ResourceMemory: "4Gi"}),
+					map[v1.ResourceName]string{v1.ResourceCPU: "4", v1.ResourceMemory: "10Gi"}).Obj(),
```
Comment (fromanirh) [inline test/integration/noderesourcetopology_test.go]: this is again a good test, because the cumulative pod requests would pass the noderesourcesfit, hence no overlap

File: test/integration/noderesourcetopology_test.go [hunk 5]
```diff
@@ -1035,7 +1331,7 @@ func checkPodSchedulingFailed(c clientset.Interface, podNamespace, podName strin
 	}
 	events, err := c.CoreV1().Events(podNamespace).List(context.TODO(), opts)
 	if err != nil {
-		klog.ErrorS(err, "cannot get events for pod %s/%s", podNamespace, podName)
+		klog.ErrorS(err, "cannot get events for pod", podNamespace, podName)
```
Comment (fromanirh) [inline test/integration/noderesourcetopology_test.go:1338]: thanks for this fix! since we are at it, we may want to fully utilize the structured format:
```
klog.ErrorS(err, "cannot get events for pod", "namespace", podNamespace, "name", podName)
```

File: test/integration/noderesourcetopology_test.go [hunk 6]
```diff
@@ -839,7 +856,286 @@ func TestTopologyMatchPlugin(t *testing.T) {
 			},
 			expectedNodes: []string{},
 		},
-	} {
+	}
+
+	// a dedicated set of tests for the scope=container flow
+	scopeEqualsContainerTests := []nrtTestUserEntry{
+		{
+			description: "[1][tier3] single container with good allocation - fit",
```
Comment (fromanirh) [inline test/integration/noderesourcetopology_test.go:868]: this is nice. My only (but not tiny) gripe is that both the expected outcome and most notably the NRT data is buried in the parsing helper, which requires back and forth from/out the helper to get the full test context.
I'm not sure how a good solution could look like though. Albeit quite uncomfortable, is not a blocker for this PR though.

Comment (fromanirh) [reply test/integration/noderesourcetopology_test.go:868]: Additionally, I think integration test should cover only tier1 (as identified in `TESTS.md` in https://github.com/kubernetes-sigs/scheduler-plugins/pull/371). Otherwise this becomes unwieldly.
- unit tests are the cheapest, so they should give the most coverage
- integration tests should focus on the most important flows (at least initially)

File: test/integration/noderesourcetopology_test.go [hunk 7]
```diff
@@ -58,6 +58,10 @@ const (
 	reasonFailedScheduling = "FailedScheduling"
 )
 
+const (
+	testPodName = "topology-aware-scheduler-pod"
```
Comment (fromanirh) [inline test/integration/noderesourcetopology_test.go:62]: "make it easier for maintain and more readable" - but also adds a hidden (and in this case tiny) interdependency between test cases :)
this is in general, in this case this seems a good change, however.

File: test/integration/noderesourcetopology_test.go [hunk 8]
```diff
@@ -73,8 +73,8 @@ type nrtTestUserEntry struct {
 	// description contains the test type and tier as described in noderesourcetopology/TESTS.md
 	// and a short description of the test itself
 	description string
-	initCntReq  []map[v1.ResourceName]string
-	cntReq      []map[v1.ResourceName]string
+	initCntReq  []map[string]string
```
Comment (fromanirh) [inline test/integration/noderesourcetopology_test.go:76]: we should probably squash this commit in the main one, especially considering we are adding this infra in this PR

**Review by fromanirh — COMMENTED** [review]
initial review. I will have a pass later to the tier1 test cases.

File: test/integration/noderesourcetopology_test.go [hunk 9]
```diff
@@ -555,6 +559,286 @@ func TestTopologyMatchPlugin(t *testing.T) {
...
+		{
+			name: "SingleNUMANodeContainerLevel: Filtering out nodes that cannot fit resources in case of Guaranteed pod with multi containers",
+			pods: []*v1.Pod{
+				withReqAndLimit(
+					withReqAndLimit(
+						st.MakePod().Namespace(ns).Name("topology-aware-scheduler-pod"),
+						map[v1.ResourceName]string{v1.ResourceCPU: "3", v1.ResourceMemory: "5Gi"}),
+					map[v1.ResourceName]string{v1.ResourceCPU: "3", v1.ResourceMemory: "5Gi"}).Obj(),
+			},
+			nodeResourceTopologies: []*topologyv1alpha1.NodeResourceTopology{
```
[the remark this answers is not in the document]
Comment (Tal-or) [reply test/integration/noderesourcetopology_test.go:630]: Done!

File: test/integration/noderesourcetopology_test.go (hunk 1 above)
Comment (Tal-or) [reply test/integration/noderesourcetopology_test.go:657]: Done!

File: test/integration/noderesourcetopology_test.go (hunk 2 above)
Comment (Tal-or) [reply test/integration/noderesourcetopology_test.go:692]: Done!

**Review by Tal-or — COMMENTED** [review]
Thank you I changed the tests accordingly and added a comment to explain the rationale behind it.

**Comment by Tal-or** [comment]
/hold cancel 
#371 merged

File: test/integration/noderesourcetopology_test.go [hunk 10]
```diff
@@ -555,6 +559,254 @@ func TestTopologyMatchPlugin(t *testing.T) {
...
+							noderesourcetopology.MakeTopologyResInfo(cpu, "4", "4"),
+							noderesourcetopology.MakeTopologyResInfo(memory, "8Gi", "8Gi"),
+						},
+					}),
+			},
+			expectedNodes: []string{"fake-node-2"},
+		},
+		// the following tests are different from the NodeResourceFit tests in the sense that
+		// they would pass the NodeResourceFit since there are enough resources for the pod to be fit
+		// at the node level, but not at the node level
```
Comment (fromanirh) [inline test/integration/noderesourcetopology_test.go:620]: `...but not at the NUMA node level`?

**Review by fromanirh — COMMENTED** [review]
LGTM
I think we can move forward

File: test/integration/noderesourcetopology_test.go (hunk 10 above)
Comment (Tal-or) [reply test/integration/noderesourcetopology_test.go:620]: thanks, I'll fix

File: test/integration/noderesourcetopology_test.go [hunk 11]
```diff
@@ -555,6 +559,254 @@ func TestTopologyMatchPlugin(t *testing.T) {
...
+			name: "SingleNUMANodeContainerLevel: Filtering out nodes that cannot fit resources in case of Guaranteed pod with multi containers",
+			pods: []*v1.Pod{
+				withReqAndLimit(
+					withReqAndLimit(
+						st.MakePod().Namespace(ns).Name("topology-aware-scheduler-pod"),
+						map[v1.ResourceName]string{v1.ResourceCPU: "3", v1.ResourceMemory: "5Gi"}),
+					map[v1.ResourceName]string{v1.ResourceCPU: "3", v1.ResourceMemory: "5Gi"}).Obj(),
+			},
+			nodeResourceTopologies: []*topologyv1alpha1.NodeResourceTopology{
+				makeNodeResourceTopology("fake-node-1", topologyv1alpha1.SingleNUMANodeContainerLevel,
```
[the remark this answers is not in the document]
Comment (swatisehgal) [reply test/integration/noderesourcetopology_test.go:631]: My take on this is that when refactoring code in a certain scope (test/integration/noderesourcetopology_test.go here) using helper functions/methods is favorable rather than using repetitive inline code for improving code readability (which is what you are doing here and is great :) ) but it is important to have a holistic view and ensure there is consistency in the code.

We shouldn't loose sight of how the file/component would look like when a new contributor starts looking at it, They shouldn't be left puzzled because of different implementation styles e.g. certain tests implemented with helper functions and others without them.

[the remark this answers is not in the document]
Comment (Tal-or) [reply test/integration/noderesourcetopology_test.go:430]: Done. Thank you :)

File: test/integration/noderesourcetopology_test.go [hunk 12]
```diff
@@ -555,6 +559,286 @@ func TestTopologyMatchPlugin(t *testing.T) {
...
+						{
+							noderesourcetopology.MakeTopologyResInfo(cpu, "4", "4"),
+							noderesourcetopology.MakeTopologyResInfo(memory, "8Gi", "8Gi"),
+						},
+					}),
+			},
+			expectedNodes: []string{},
+		},
+		{
+			name: "Negative: SingleNUMANodeContainerLevel: Cannot fit resources in case of Guaranteed pod with multi containers",
```
[the remark this answers is not in the document]
Comment (Tal-or) [reply test/integration/noderesourcetopology_test.go:596]: The ```expectedNode``` list contains a list of nodes that we expect the tested pod to be landing on one of them.
If the list is empty (like in this case) it means we expect the pod to keep in ```Pending``` state.

By taking a look at the NRT object of ```fake-node-2``` it seems the pod should be landing there but yet the test passed (i.e. the pod kept in ```Pending```).
What actually happens is that the pod landed on ```fake-node-2``` but there was a bug in the testing logic (which I just fixed) that didn't detect pending pods accurately.

I also changed the NRT object of ```fake-node-2``` values so now the tested pod won't be able to land there.

Thank you for catching this!

File: test/integration/noderesourcetopology_test.go [hunk 13]
```diff
@@ -932,7 +931,7 @@ func getNodeName(ctx context.Context, c clientset.Interface, podNamespace, podNa
 func createNodeResourceTopologies(ctx context.Context, topologyClient *versioned.Clientset, noderesourcetopologies []*topologyv1alpha1.NodeResourceTopology) error {
 	for _, nrt := range noderesourcetopologies {
 		_, err := topologyClient.TopologyV1alpha1().NodeResourceTopologies().Create(ctx, nrt, metav1.CreateOptions{})
-		if err != nil && !errors.IsAlreadyExists(err) {
+		if err != nil {
```
Comment (fromanirh) [inline test/integration/noderesourcetopology_test.go:934]: good catch!

**Review by fromanirh — COMMENTED** [review]
LGTM (non-binding)

File: test/integration/noderesourcetopology_test.go (hunk 13 above)
Comment (Tal-or) [reply test/integration/noderesourcetopology_test.go:934]: This was not a bug. 
It's just wrong.
There is no reason to have the NRT object exist before the test start.
The environment should always stays clean and if it's not we should report it back.

Comment (fromanirh) [reply test/integration/noderesourcetopology_test.go:934]: sure, agree with your assessment

**Review by Huang-Wei — COMMENTED** [review]
Some comments on coding style.

File: test/integration/noderesourcetopology_test.go [hunk 14]
```diff
@@ -607,7 +624,234 @@ func TestTopologyMatchPlugin(t *testing.T) {
...
+		{
+			description: "[32][tier1] multi init containers with over memory allocation - not fit",
+			initCntReq: []map[string]string{
+				{cpu: "20", memory: "50G"},
+				{cpu: "40", memory: "50G"},
+				{cpu: "3", memory: "50G"},
+			},
+			cntReq: []map[string]string{
+				{cpu: "20", memory: "40G"},
+				{cpu: "2", memory: "6G"},
```
[the remark this answers is not in the document]
Comment (swatisehgal) [reply test/integration/noderesourcetopology_test.go:850]: Also noticed the use of "G" in `60G` instead of `60Gi` here and in all the entries of `scopeEqualsContainerTests`.

G means 5 Gigabytes while Gi means Gibibytes. , For example:
5 G = 5000000 KB / 5000 MB
5 Gi = 5368709.12 KB / 5368.70 MB

I would prefer the use of units comparable to the ones we have in NRT object. If this is intentional however, please make it very explicit because this is very easy to miss.

File: test/integration/noderesourcetopology_test.go [hunk 15]
```diff
@@ -607,7 +624,234 @@ func TestTopologyMatchPlugin(t *testing.T) {
 			},
 			expectedNodes: []string{},
 		},
-	} {
+	}
+
+	// a dedicated set of tests for the scope=container flow
+	scopeEqualsContainerTests := []nrtTestUserEntry{
+		{
+			description: "[4][tier1] multi containers with good devices and hugepages allocation, spread across NUMAs - fit",
```
[the remark this answers is not in the document]
Comment (swatisehgal) [reply test/integration/noderesourcetopology_test.go:635]: After reviewing all the test cases, I understand what you meant here and the description makes sense. Feel free to ignore the comment above.

File: test/integration/noderesourcetopology_test.go [hunk 16]
```diff
@@ -639,27 +947,38 @@ func cleanupNodeResourceTopologies(ctx context.Context, topologyClient *versione
...
 		Resources: v1.ResourceRequirements{
 			Requests: res,
 			Limits:   res,
 		},
 	})
+	return cList
+}
+
+// initWithReqAndLimit adds a new init container to the inner pod with given resource map.
+func initWithReqAndLimit(p *st.PodWrapper, resMap map[string]string) *st.PodWrapper {
```
[the remark this answers is not in the document]
Comment (Tal-or) [reply test/integration/noderesourcetopology_test.go:967]: good idea. we can even name it ```withLimits``` which force ```Requests``` to be equal anyway.

File: test/integration/noderesourcetopology_test.go [hunk 17]
```diff
@@ -700,10 +944,10 @@ func cleanupNodeResourceTopologies(ctx context.Context, topologyClient *versione
 	}
 }
 
-func withReqAndLimitHelper(resMap map[v1.ResourceName]string, cList []v1.Container, prefix string) []v1.Container {
+func withReqAndLimitHelper(resMap map[string]string, cList []v1.Container, prefix string) []v1.Container {
```
[the remark this answers is not in the document]
Comment (Tal-or) [reply test/integration/noderesourcetopology_test.go:950]: This way we can use a shorter and more readable version for initializing the tests:
using:

```
cntReq: []map[string]string{
				{cpu: "20", memory: "40G"},
				{cpu: "20", memory: "40G"},
				{cpu: "20", memory: "10G"},
			},
```

instead of:

```
cntReq: []map[v1.ResourceName]string{
				{v1.ResourceCPU: "20", v1.ResourceMemory: "40G"},
				{v1.ResourceCPU: "20", v1.ResourceMemory: "40G"},
				{v1.ResourceCPU: "20", v1.ResourceMemory: "10G"},
			},
```
Eventually,  ```withReqAndLimitHelper``` gets ```cntReq```

File: test/integration/noderesourcetopology_test.go [hunk 18]
```diff
@@ -795,3 +1039,87 @@ func makeNodeResourceTopology(name string, tmp topologyv1alpha1.TopologyManagerP
...
+							noderesourcetopology.MakeTopologyResInfo(memory, "64Gi", "64Gi"),
+							noderesourcetopology.MakeTopologyResInfo(hugepages2Mi, "512Mi", "512Mi"),
+							noderesourcetopology.MakeTopologyResInfo(nicResourceName, "32", "32"),
+						},
+					},
+				},
+			},
+			{
+				// we set all available resource as 0
+				// because we want all these tests to be running against fake-node-1
```
[the remark this answers is not in the document]
Comment (Tal-or) [reply test/integration/noderesourcetopology_test.go:1086]: ```NodeResourceFit``` won't filter ```fake-node-2``` because it has enough resources. 
Next, if we won't create ```noderesourcetopology``` object for ```fake-node-2``` with zero resources, it won't be filtered by our plugin either.
IOW the node would be marked as a suitable one and it will affect the tests results`

File: test/integration/noderesourcetopology_test.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (Tal-or) [reply test/integration/noderesourcetopology_test.go:850]: No. Actually, it should be the init container which is failed. I fixed that.
About the ```G``` notation, I used that because it was easier to calculate the values needed for the test.
But if the NRT objects are using ```Gi``` it won't make any difference.
I'll change that to ```Gi``` for consistency.

File: test/integration/noderesourcetopology_test.go [hunk 19]
```diff
@@ -146,386 +171,185 @@ func TestTopologyMatchPlugin(t *testing.T) {
...
-							},
-						},
-						topologyv1alpha1.Zone{
-							Name: "node-1",
-							Type: "Node",
-							Resources: topologyv1alpha1.ResourceInfoList{
-								noderesourcetopology.MakeTopologyResInfo(cpu, "4", "4"),
-								noderesourcetopology.MakeTopologyResInfo(memory, "50Gi", "50Gi"),
-							},
+				makeNodeResourceTopology("fake-node-1", topologyv1alpha1.SingleNUMANodeContainerLevel,
```
[the remark this answers is not in the document]
Comment (Tal-or) [reply test/integration/noderesourcetopology_test.go]: Very nice. Done

File: test/integration/noderesourcetopology_test.go [hunk 20]
```diff
@@ -555,6 +555,254 @@ func TestTopologyMatchPlugin(t *testing.T) {
...
+					}),
+				makeNodeResourceTopology("fake-node-2", topologyv1alpha1.SingleNUMANodeContainerLevel,
+					[]topologyv1alpha1.ResourceInfoList{
+						{
+							noderesourcetopology.MakeTopologyResInfo(cpu, "4", "4"),
+							noderesourcetopology.MakeTopologyResInfo(memory, "8Gi", "8Gi"),
+						},
+						{
+							noderesourcetopology.MakeTopologyResInfo(cpu, "4", "4"),
+							noderesourcetopology.MakeTopologyResInfo(memory, "8Gi", "4Gi"),
```
[the remark this answers is not in the document]
Comment (Tal-or) [reply test/integration/noderesourcetopology_test.go]: Done

File: test/integration/noderesourcetopology_test.go [hunk 21]
```diff
@@ -58,6 +60,24 @@ var (
...
 
+// should be filled by the user
+type nrtTestUserEntry struct {
+	// description contains the test type and tier as described in noderesourcetopology/TESTS.md
+	// and a short description of the test itself
+	description string
+	initCntReq  []map[string]string
+	cntReq      []map[string]string
+	statusErr   string
+	// this testing batch is going to br run against the same node and NRT objects, hence we're not specifying them.
```
[the remark this answers is not in the document]
Comment (Tal-or) [reply test/integration/noderesourcetopology_test.go]: Thanks. Done

File: test/integration/noderesourcetopology_test.go [hunk 22]
```diff
@@ -795,3 +1039,87 @@ func makeNodeResourceTopology(name string, tmp topologyv1alpha1.TopologyManagerP
...
+	for i, e := range entries {
+		p := st.MakePod().Name(fmt.Sprintf("%s-%d", "topology-aware-scheduler-pod", i+1)).Namespace(ns)
+		for _, req := range e.initCntReq {
+			p = initWithReqAndLimit(p, req)
+		}
+
+		for _, req := range e.cntReq {
+			p = withReqAndLimit(p, req)
+		}
+		nodeTopologies := []*topologyv1alpha1.NodeResourceTopology{
```
[the remark this answers is not in the document]
Comment (Tal-or) [reply test/integration/noderesourcetopology_test.go:1083]: Thanks :)

File: test/integration/noderesourcetopology_test.go (hunk 17 above)
[the remark this answers is not in the document]
Comment (swatisehgal) [reply test/integration/noderesourcetopology_test.go]: Ack, Thanks for the clarification.

File: test/integration/noderesourcetopology_test.go (hunk 18 above)
[the remark this answers is not in the document]
Comment (swatisehgal) [reply test/integration/noderesourcetopology_test.go]: Yeah, I recall now a scenario that was pointed out to us earlier where NRT objects did not exist. Thanks for explaining.

**Comment by swatisehgal** [comment]
/lgtm

Thanks @Tal-or for addressing all my review comments. I am happy with the PR now :) Looking nice!

**Comment by Huang-Wei** [comment]
/approve

Thanks @Tal-or !
